Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -99,65 +100,146 @@ namespace Terrain
+            );
+            m_areaData.m_heightmapImage = AZ::RPI::AttachmentImage::Create(*imagePool.get(), imageDescriptor, AZ::Name("TerrainHeightmap"), nullptr, nullptr);
+            AZ_Error(TerrainFPName, m_areaData.m_heightmapImage, "Failed to initialize the heightmap image!");
+        }
+
+        AZStd::vector<uint16_t> pixels;
+        pixels.reserve(width * height);
+        const uint32_t pixelDataSize = width * height * sizeof(uint16_t);
 
-        // Create heightmap image data
         {
-            m_areaData.m_propertiesDirty = true;
+            // Block other threads from accessing the surface data bus while we are in GetHeightFromFloats (which may call into the SurfaceData bus).
+            // We lock our surface data mutex *before* checking / setting "isRequestInProgress" so that we prevent race conditions
+            // that create false detection of cyclic dependencies when multiple requests occur on different threads simultaneously.
+            // (One case where this was previously able to occur was in rapid updating of the Preview widget on the
+            // GradientSurfaceDataComponent in the Editor when moving the threshold sliders back and forth rapidly)
 
-            AZ::RHI::Size imageSize;
-            imageSize.m_width = width;
-            imageSize.m_height = height;
+            auto& surfaceDataContext = SurfaceData::SurfaceDataSystemRequestBus::GetOrCreateContext(false);
+            typename SurfaceData::SurfaceDataSystemRequestBus::Context::DispatchLockGuard scopeLock(surfaceDataContext.m_contextMutex);
 
-            AZStd::vector<uint16_t> uint16Heights;
-            uint16Heights.reserve(heightData.size());
-            for (float sampleHeight : heightData)
+            for (uint32_t y = 0; y < height; y++)
             {
-                float clampedSample = AZ::GetClamp(sampleHeight, 0.0f, 1.0f);
-                constexpr uint16_t MaxUint16 = 0xFFFF;
-                uint16Heights.push_back(aznumeric_cast<uint16_t>(clampedSample * MaxUint16));
+                for (uint32_t x = 0; x < width; x++)
+                {
+                    bool terrainExists = true;
+                    float terrainHeight = 0.0f;
+                    AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+                        terrainHeight, &AzFramework::Terrain::TerrainDataRequests::GetHeightFromFloats,
+                        (x * queryResolution) + m_dirtyRegion.GetMin().GetX(),
+                        (y * queryResolution) + m_dirtyRegion.GetMin().GetY(),
+                        AzFramework::Terrain::TerrainDataRequests::Sampler::EXACT,
+                        &terrainExists);
+
+                    float clampedHeight = AZ::GetClamp((terrainHeight - worldBounds.GetMin().GetZ()) / worldBounds.GetExtents().GetZ(), 0.0f, 1.0f);
+                    constexpr uint16_t MaxUint16 = 0xFFFF;
+
+                    pixels.push_back(aznumeric_cast<uint16_t>(clampedHeight * MaxUint16));
```

Comment:
This could use AZStd::numeric_limits<uint16_t>::max() directly instead of defining the value.

Reply:
Good call. Looking at this again made me realize we really want to round the values to the nearest uint16_t value, not floor it, so I need to adjust this slightly.